File: App/uni_modules/tuikit-atomic-x/utssdk/app-android/kotlin/CallRenderView.kt
package uts.sdk.modules.atomicx.kotlin

import android.content.Context
import android.util.AttributeSet
import android.util.Log
import androidx.constraintlayout.widget.ConstraintLayout
import com.google.gson.Gson
import io.dcloud.uts.console
import io.trtc.tuikit.atomicxcore.api.device.NetworkQuality
import io.trtc.tuikit.atomicxcore.api.view.CallCoreView
import io.trtc.tuikit.atomicxcore.api.view.CallLayoutTemplate
import io.trtc.tuikit.atomicxcore.api.view.VolumeLevel

private const val TAG = "UTS-CallRenderView: "

class CallRenderView(
  context: Context,
  attrs: AttributeSet? = null,
) : ConstraintLayout(context, attrs) {
  private var nativeCallCoreView: CallCoreView? = null

  private var cachedLayoutTemplate: CallLayoutTemplate = CallLayoutTemplate.Grid
  private var cachedWaitingAnimation: String = ""
  private var cachedVolumeLevelIcons: Map<VolumeLevel, String> = mapOf()
  private var cachedNetworkQualityIcons: Map<NetworkQuality, String> = mapOf()
  private var cachedParticipantAvatars: Map<String, String> = mapOf()
  private var isNativeViewInitialized = false

  private val gson = Gson()

  override fun onAttachedToWindow() {
    super.onAttachedToWindow()
    Log.w(TAG, "onAttachedToWindow")
    console.log("$TAG onAttachedToWindow")

    initializeView()
  }

  override fun onDetachedFromWindow() {
    super.onDetachedFromWindow()
    Log.w(TAG, "onDetachedFromWindow")
    console.log("$TAG onDetachedFromWindow")

    nativeCallCoreView = null
    isNativeViewInitialized = false
  }

  fun setLayoutTemplate(layoutTemplate: Any?) {
    console.warn("$TAG setLayoutTemplate:", layoutTemplate)
    Logger.i(TAG + "setLayoutTemplate: $layoutTemplate")

    if (layoutTemplate !is String) {
      return
    }

    cachedLayoutTemplate =
      when (layoutTemplate) {
        "Float" -> CallLayoutTemplate.Float
        "Grid" -> CallLayoutTemplate.Grid
        "Pip" -> CallLayoutTemplate.Pip
        else -> CallLayoutTemplate.Grid
      }

    nativeCallCoreView?.setLayoutTemplate(cachedLayoutTemplate)
  }

  fun setWaitingAnimation(animationPath: Any?) {
    console.warn("$TAG setWaitingAnimation:", animationPath)
    Logger.i(TAG + "setWaitingAnimation: $animationPath")

    if (animationPath !is String) {
      return
    }
    cachedWaitingAnimation = animationPath
    nativeCallCoreView?.setWaitingAnimation(cachedWaitingAnimation)
  }

  fun setVolumeLevelIcons(icons: Any?) {
    console.warn("$TAG setVolumeLevelIcons:", icons)
    Logger.i(TAG + "setVolumeLevelIcons: $icons")

    if (icons !is String) {
      return
    }

    try {
      val mapData = gson.fromJson<MutableMap<String, String?>>(icons.toString(), MutableMap::class.java)
      if (mapData == null) {
        console.error("$TAG Failed to convert icons to Map")
        Logger.e(TAG + "Failed to convert icons to Map")
        return
      }

      val volumeIconMap = mutableMapOf<VolumeLevel, String>()
      mapData.forEach { (key, value) ->
        if (value is String) {
          val volumeLevel =
            when (key) {
              "Mute" -> VolumeLevel.Mute
              "Low" -> VolumeLevel.Low
              "Medium" -> VolumeLevel.Medium
              "High" -> VolumeLevel.High
              "Peak" -> VolumeLevel.Peak
              else -> null
            }
          if (volumeLevel != null) {
            volumeIconMap[volumeLevel] = value
          }
        }
      }

      cachedVolumeLevelIcons = volumeIconMap
      nativeCallCoreView?.setVolumeLevelIcons(cachedVolumeLevelIcons)
    } catch (e: Exception) {
      console.error("$TAG setVolumeLevelIcons failed:", e.message)
      Logger.e(TAG + "setVolumeLevelIcons failed: ${e.message}")
    }
  }

  fun setNetworkQualityIcons(icons: Any?) {
    console.warn("$TAG setNetworkQualityIcons:", icons)
    Logger.i(TAG + "setNetworkQualityIcons: $icons")

    if (icons !is String) {
      return
    }

    try {
      val mapData = gson.fromJson<MutableMap<String, String?>>(icons.toString(), MutableMap::class.java)
      if (mapData == null) {
        console.error("$TAG Failed to convert icons to Map")
        Logger.e(TAG + "Failed to convert icons to Map")
        return
      }

      val networkIconMap = mutableMapOf<NetworkQuality, String>()
      mapData.forEach { (key, value) ->
        if (value is String) {
          val networkQuality =
            when (key) {
              "UNKNOWN" -> NetworkQuality.UNKNOWN
              "EXCELLENT" -> NetworkQuality.EXCELLENT
              "GOOD" -> NetworkQuality.GOOD
              "POOR" -> NetworkQuality.POOR
              "BAD" -> NetworkQuality.BAD
              "VERY_BAD" -> NetworkQuality.VERY_BAD
              "DOWN" -> NetworkQuality.DOWN
              else -> null
            }
          if (networkQuality != null) {
            networkIconMap[networkQuality] = value
          }
        }
      }

      cachedNetworkQualityIcons = networkIconMap
      nativeCallCoreView?.setNetworkQualityIcons(cachedNetworkQualityIcons)
    } catch (e: Exception) {
      console.error("$TAG setNetworkQualityIcons failed:", e.message)
      Logger.e(TAG + "setNetworkQualityIcons failed: ${e.message}")
    }
  }

  fun setParticipantAvatars(avatars: Any) {
    console.warn("$TAG setParticipantAvatars: ", avatars)
    Logger.i(TAG + "setParticipantAvatars: $avatars")

    if (avatars !is String) {
      return
    }

    try {
      val mapData = gson.fromJson<MutableMap<String, String?>>(avatars.toString(), MutableMap::class.java)
      if (mapData == null) {
        console.error("$TAG Failed to convert avatars to Map")
        Logger.e(TAG + "Failed to convert avatars to Map")
        return
      }

      val avatarMap = mutableMapOf<String, String>()

      mapData.forEach { (key, value) ->
        if (value is String) {
          avatarMap[key] = value
        }
      }

      console.log("$TAG avatar:$avatarMap")
      Logger.i(TAG + "avatar: $avatarMap")

      cachedParticipantAvatars = avatarMap
      nativeCallCoreView?.setParticipantAvatars(cachedParticipantAvatars)
    } catch (e: Exception) {
      console.error("$TAG setParticipantAvatars failed:", e.message)
      Logger.e(TAG + "setParticipantAvatars failed: ${e.message}")
      e.printStackTrace()
    }
  }

  private fun initializeView() {
    if (isNativeViewInitialized) {
      console.warn("$TAG initializeView: already initialized")
      Logger.w(TAG + "initializeView: already initialized")
      return
    }
    console.warn("$TAG initializeView")
    Logger.i(TAG + "initializeView")
    removeAllViews()

    val nativeView = CallCoreView(context, null, 0)
    nativeView.setLayoutTemplate(cachedLayoutTemplate)
    if (cachedWaitingAnimation.isNotEmpty()) {
      nativeView.setWaitingAnimation(cachedWaitingAnimation)
    }

    if (cachedVolumeLevelIcons.isNotEmpty()) {
      nativeView.setVolumeLevelIcons(cachedVolumeLevelIcons)
    }

    if (cachedNetworkQualityIcons.isNotEmpty()) {
      nativeView.setNetworkQualityIcons(cachedNetworkQualityIcons)
    }

    if (cachedParticipantAvatars.isNotEmpty()) {
      nativeView.setParticipantAvatars(cachedParticipantAvatars)
    }

    val layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
    addView(nativeView, layoutParams)
    nativeCallCoreView = nativeView
    isNativeViewInitialized = true
  }
}

File: App/uni_modules/tuikit-atomic-x/utssdk/app-ios/swift/CallRenderView.swift
import AtomicXCore
import DCloudUTSFoundation
import RTCRoomEngine
import UIKit

public class CallRenderView: UIView {
    private static let TAG = "iOS-CallRenderView: "
    
    private var nativeCallCoreView: AtomicXCore.CallCoreView?
    
    private var cachedLayoutTemplate: AtomicXCore.CallLayoutTemplate = .grid
    private var cachedWaitingAnimation: String = ""
    private var cachedVolumeLevelIcons: [AtomicXCore.VolumeLevel: String] = [:]
    private var cachedNetworkQualityIcons: [NetworkQuality: String] = [:]
    private var cachedParticipantAvatars: [String: String] = [:]
    private var isNativeViewInitialized = false
    
    override public init(frame: CGRect = .zero) {
        super.init(frame: frame)
        commonInit()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        commonInit()
    }
    
    private func commonInit() {
        console.log(Self.TAG, "init CallRenderView wrapper")
    }
    
    public override func didMoveToWindow() {
        super.didMoveToWindow()
        if window != nil {
            console.log(Self.TAG, "didMoveToWindow - view attached")
            initializeView()
        } else {
            console.log(Self.TAG, "didMoveToWindow - view detached")
        }
    }
    
    public override func removeFromSuperview() {
        console.log(Self.TAG, "removeFromSuperview")
        super.removeFromSuperview()
        nativeCallCoreView?.removeFromSuperview()
        nativeCallCoreView = nil
        isNativeViewInitialized = false
    }
    
    public override func layoutSubviews() {
        super.layoutSubviews()
        nativeCallCoreView?.frame = bounds
    }
    
    public func setLayoutTemplate(_ layoutTemplate: Any) {
        console.warn(Self.TAG, "setLayoutTemplate:", layoutTemplate)
        
        guard let layoutTemplateStr = layoutTemplate as? String else {
            console.error(Self.TAG, "setLayoutTemplate: layoutTemplate is not String")
            return
        }
        
        switch layoutTemplateStr {
        case "Float", "float":
            cachedLayoutTemplate = .float
        case "Grid", "grid":
            cachedLayoutTemplate = .grid
        case "Pip", "pip":
            cachedLayoutTemplate = .pip
        default:
            console.warn(Self.TAG, "Unknown layoutTemplate:", layoutTemplateStr, ", using Grid")
            cachedLayoutTemplate = .grid
        }
        
        nativeCallCoreView?.setLayoutTemplate(cachedLayoutTemplate)
    }
    
    public func setWaitingAnimation(_ animationPath: Any) {
        console.warn(Self.TAG, "setWaitingAnimation:", animationPath)
        
        guard let pathStr = animationPath as? String else {
            console.error(Self.TAG, "setWaitingAnimation: animationPath is not String")
            return
        }
        
        cachedWaitingAnimation = pathStr
        
        if !cachedWaitingAnimation.isEmpty {
            nativeCallCoreView?.setWaitingAnimation(path: cachedWaitingAnimation)
        }
    }
    
    public func setVolumeLevelIcons(_ icons: Any) {
        console.warn(Self.TAG, "setVolumeLevelIcons:", icons)
        
        guard let iconsStr = icons as? String,
              let data = iconsStr.data(using: .utf8),
              let iconsDict = try? JSONSerialization.jsonObject(with: data) as? [String: String] else {
            console.error(Self.TAG, "setVolumeLevelIcons: failed to parse icons JSON, actual type:", String(describing: type(of: icons)))
            return
        }
        
        var volumeIconMap: [AtomicXCore.VolumeLevel: String] = [:]
        for (key, value) in iconsDict {
            let volumeLevel: AtomicXCore.VolumeLevel?
            switch key {
            case "Mute", "mute":
                volumeLevel = .mute
            case "Low", "low":
                volumeLevel = .low
            case "Medium", "medium":
                volumeLevel = .medium
            case "High", "high":
                volumeLevel = .high
            case "Peak", "peak":
                volumeLevel = .peak
            default:
                volumeLevel = nil
            }
            
            if let level = volumeLevel {
                volumeIconMap[level] = value
            }
        }
        
        cachedVolumeLevelIcons = volumeIconMap
        
        if !cachedVolumeLevelIcons.isEmpty {
            nativeCallCoreView?.setVolumeLevelIcons(icons: cachedVolumeLevelIcons)
        }
    }
    
    public func setNetworkQualityIcons(_ icons: Any) {
        console.warn(Self.TAG, "setNetworkQualityIcons:", icons)
        
        guard let iconsStr = icons as? String,
              let data = iconsStr.data(using: .utf8),
              let iconsDict = try? JSONSerialization.jsonObject(with: data) as? [String: String] else {
            console.error(Self.TAG, "setNetworkQualityIcons: failed to parse icons JSON, actual type:", String(describing: type(of: icons)))
            return
        }
        
        var networkIconMap: [NetworkQuality: String] = [:]
        for (key, value) in iconsDict {
            let networkQuality: NetworkQuality?
            switch key {
            case "UNKNOWN", "unknown":
                networkQuality = .unknown
            case "EXCELLENT", "excellent":
                networkQuality = .excellent
            case "GOOD", "good":
                networkQuality = .good
            case "POOR", "poor":
                networkQuality = .poor
            case "BAD", "bad":
                networkQuality = .bad
            case "VERY_BAD", "veryBad", "very_bad":
                networkQuality = .veryBad
            case "DOWN", "down":
                networkQuality = .down
            default:
                networkQuality = nil
            }
            
            if let quality = networkQuality {
                networkIconMap[quality] = value
            }
        }
        
        cachedNetworkQualityIcons = networkIconMap
        
        if !cachedNetworkQualityIcons.isEmpty {
            nativeCallCoreView?.setNetworkQualityIcons(icons: cachedNetworkQualityIcons)
        }
    }
    
    public func setParticipantAvatars(_ avatars: Any) {
        console.warn(Self.TAG, "setParticipantAvatars:", avatars)
        
        guard let avatarsStr = avatars as? String,
              let data = avatarsStr.data(using: .utf8),
              let avatarsDict = try? JSONSerialization.jsonObject(with: data) as? [String: String] else {
            console.error(Self.TAG, "setParticipantAvatars: failed to parse avatars JSON, actual type:", String(describing: type(of: avatars)))
            return
        }
        
        cachedParticipantAvatars = avatarsDict
        
        if !cachedParticipantAvatars.isEmpty {
            nativeCallCoreView?.setParticipantAvatars(avatars: cachedParticipantAvatars)
        }
    }
    
    public func initializeView() {
        console.warn(Self.TAG, "initializeView")
        
        if isNativeViewInitialized {
            console.warn(Self.TAG, "initializeView: already initialized")
            return
        }
        
        subviews.forEach { $0.removeFromSuperview() }
        
        let nativeView = AtomicXCore.CallCoreView(frame: bounds)
        
        nativeView.setLayoutTemplate(cachedLayoutTemplate)
        
        if !cachedWaitingAnimation.isEmpty {
            nativeView.setWaitingAnimation(path: cachedWaitingAnimation)
        }
        
        if !cachedVolumeLevelIcons.isEmpty {
            nativeView.setVolumeLevelIcons(icons: cachedVolumeLevelIcons)
        }
        
        if !cachedNetworkQualityIcons.isEmpty {
            nativeView.setNetworkQualityIcons(icons: cachedNetworkQualityIcons)
        }
        
        if !cachedParticipantAvatars.isEmpty {
            nativeView.setParticipantAvatars(avatars: cachedParticipantAvatars)
        }
        
        nativeView.translatesAutoresizingMaskIntoConstraints = false
        addSubview(nativeView)
        
        NSLayoutConstraint.activate([
            nativeView.leadingAnchor.constraint(equalTo: leadingAnchor),
            nativeView.trailingAnchor.constraint(equalTo: trailingAnchor),
            nativeView.topAnchor.constraint(equalTo: topAnchor),
            nativeView.bottomAnchor.constraint(equalTo: bottomAnchor)
        ])
        
        nativeCallCoreView = nativeView
        isNativeViewInitialized = true
        
        console.log(Self.TAG, "initializeView success")
    }
}
